### PR TITLE
Rig conversion

### PR DESCRIPTION
**Pull request by sara-es**
Converted the Rig algorithm hilbert into a PACMAN algorithm hilbert_placer with utility object hilbert_state, and included a test isomorph_check that can be run to check if an isomorphism exists between placements by two different algorithms.

### CODE REVIEW

File: pacman/operations/rigged_algorithms/hilbert_state.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+            :param ypos: the y coordinate on the generated curve
+            :param xchange: the change in x coordinate on the generated curve
+            :param ychange: the change in y coordinate on the generated curve
+            :type xpos: int
+            :type ypos: int
+            :type xchange: int
+            :type ychange: int
+    """
+    def __init__(self, xpos=0, ypos=0, xchange=1, ychange=0):
+        self.x_pos, self.y_pos, self.change_x, self.change_y = \
```
Comment (alan-stokes) [inline pacman/operations/rigged_algorithms/hilbert_state.py]: avoid \ 
likely better breaking this allocation up into 4 lines. yet again, readability vs efficiency.

File: pacman/operations/rigged_algorithms/isomorph_check.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+vertices by two separate placement algorithms. One of the algorithms must
+output to memory placements_copy and
+<param_type>MemoryPlacements2</param_type> in the algorithms_metadata.xml."""
+
+import os
+import logging
+
+
+class IsomorphicChecker(object):
+
```
Comment (alan-stokes) [inline pacman/operations/rigged_algorithms/isomorph_check.py]: same thing about class saying what it does, call having doc string saying types.

File: pacman/operations/rigged_algorithms/isomorph_check.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    def check(self, placements, placements_copy):
+        chip_vertices = set()
+        chip_vertices_copy = set()
+        chips = set()
+        correct = True
+
+        # create a list of processors with assigned vertices
+        for x, y, p in placements.get_placed_processors():
+            chips.add((x, y))
+        # create sets of the vertices on each processor and compare them
```
Comment (alan-stokes) [inline pacman/operations/rigged_algorithms/isomorph_check.py]: new line between code and comment

File: pacman/operations/rigged_algorithms/isomorph_check.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        for (x, y) in chips:
+            for p in range(0, 18):
+                if placements.is_processor_occupied(x, y, p):
```
Comment (alan-stokes) [inline pacman/operations/rigged_algorithms/isomorph_check.py]: comment

File: pacman/operations/rigged_algorithms/isomorph_check.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+                    chip_vertices.add(placements.get_vertex_on_processor(
+                        x, y, p))
+                if placements_copy.is_processor_occupied(x, y, p):
```
Comment (alan-stokes) [inline pacman/operations/rigged_algorithms/isomorph_check.py]: comment

File: pacman/operations/rigged_algorithms/isomorph_check.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+                    chip_vertices_copy.add(
+                        placements_copy.get_vertex_on_processor(x, y, p))
+        if chip_vertices != chip_vertices_copy:
```
Comment (alan-stokes) [inline pacman/operations/rigged_algorithms/isomorph_check.py]: comment

File: pacman/operations/rigged_algorithms/isomorph_check.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+            correct = False
+        print chips
+        return correct
+
+    def __call__(self, report_folder, placements,
```
Comment (alan-stokes) [inline pacman/operations/rigged_algorithms/isomorph_check.py]: style states to put call at top of script, not at bottom

File: pacman/operations/rigged_algorithms/isomorph_check.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+                 placements_copy):
```
Comment (alan-stokes) [inline pacman/operations/rigged_algorithms/isomorph_check.py]: placements_copy can be on same line as rest

File: pacman/operations/rigged_algorithms/isomorph_check.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        file_name = os.path.join(report_folder, "placement_isomorph.rpt")
+        f = None
```
Comment (alan-stokes) [inline pacman/operations/rigged_algorithms/isomorph_check.py]: f?! output?

File: pacman/operations/rigged_algorithms/isomorph_check.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        try:
+            f = open(file_name, "w")
+        except IOError:
+            logging.getLogger().error("Generate_isomorph_report: Can't open "
```
Comment (alan-stokes) [inline pacman/operations/rigged_algorithms/isomorph_check.py]: push string onto new line for readability

File: pacman/operations/rigged_algorithms/basic_test.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+from pacman.utilities.algorithm_utilities import placer_algorithm_utilities
+from pacman.model.placements import Placement, Placements
+from pacman.utilities.utility_objs import ResourceTracker
+
+# general imports
+import logging
+logger = logging.getLogger(__name__)
+
+
+class BasicPlacerTest(object):
```
Comment (alan-stokes) [inline pacman/operations/rigged_algorithms/basic_test.py]: what does this class do exactly? it looks like the basic placer but labelled test. if so, couldnt it just be removed?

File: pacman/operations/algorithms_metadata.xml [hunk 12]
```diff
@@ -139,6 +139,73 @@
...
+            <parameter>
+                <param_name>report_folder</param_name>
+                <param_type>ReportFolder</param_type>
+            </parameter>
+            <parameter>
+                <param_name>placements</param_name>
+                <param_type>MemoryPlacements</param_type>
+            </parameter>
+            <parameter>
+                <param_name>placements_copy</param_name>
```
Comment (alan-stokes) [inline pacman/operations/algorithms_metadata.xml]: placements_copy vs other_algorithm_placements?

**Comment by alan-stokes** [comment]
worth noting that the isomophic checker requires changes to the xml to operate. but given the risk of overriding two placers outputs with each other. acceptable interface. Might want to document this in the checkers python docs.

File: pacman/operations/rigged_algorithms/hilbert_placer.py [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+        # placements_copy = Placements()
+        placements = Placements()
+        vertices = \
+            placer_algorithm_utilities.sort_vertices_by_known_constraints(
+                machine_graph.vertices)
+
+        # Iterate over vertices and generate placements
+        progress = ProgressBar(machine_graph.n_vertices,
+                               "Placing graph vertices")
+        resource_tracker = ResourceTracker(machine,
```
Comment (alan-stokes) [inline pacman/operations/rigged_algorithms/hilbert_placer.py]: push params onto new line. readability

File: pacman/operations/rigged_algorithms/hilbert_placer.py [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+
+    def _place_vertex(self, vertex, resource_tracker, machine, placements,
+                      vertices_on_same_chip):
+
+        vertices = vertices_on_same_chip[vertex]
+        chips = self._generate_hilbert_chips(machine)
+
+        if len(vertices) > 1:
+            assigned_values = \
+                resource_tracker.allocate_constrained_group_resources([
```
Comment (alan-stokes) [inline pacman/operations/rigged_algorithms/hilbert_placer.py]: break param build into separate line. very difficult to read as is

File: pacman/operations/rigged_algorithms/hilbert_placer.py [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+            for (x, y, p, _, _), vert in zip(assigned_values, vertices):
+                placement = Placement(vert, x, y, p)
+                placements.add_placement(placement)
+        else:
+            (x, y, p, _, _) = resource_tracker.allocate_constrained_resources(
+                vertex.resources_required, vertex.constraints, chips)
+            placement = Placement(vertex, x, y, p)
+            placements.add_placement(placement)
+
+        return vertices
```
Comment (alan-stokes) [inline pacman/operations/rigged_algorithms/hilbert_placer.py]: comment for the file. order of functions would be easier to read if they are ordered downwards as they are used. hilbert_curve should be at bottom. _generate_hilbert_chips would be 3rd etc.
this is just a style comment though

File: pacman/operations/rigged_algorithms/basic_test.py (hunk 11 above)
Comment (sara-es) [inline pacman/operations/rigged_algorithms/basic_test.py]: Was used to test if the isomorphic checker worked without messing with the basic placer. Can be removed

File: pacman/operations/rigged_algorithms/hilbert_placer.py [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+        # position
+        if state is None:
+            state = HilbertState()
+            yield state.x_pos, state.y_pos
+
+        if level <= 0:
+            return
+
+        # Turn left
+        state.change_x, state.change_y = state.change_y * -angle, \
```
Comment (sara-es) [inline pacman/operations/rigged_algorithms/hilbert_placer.py]: \ must be used regardless

Comment (alan-stokes) [inline pacman/operations/rigged_algorithms/hilbert_placer.py]: nah, and if so, you could split the lines into 2, which is just as easy to read. You gain no performance increase doing both in 1 line. 

state.change_x =  state.change_y * -angle
state,change_y = .... whatever this was

File: pacman/operations/rigged_algorithms/hilbert_placer.py (hunk 15 above)
Comment (sara-es) [inline pacman/operations/rigged_algorithms/hilbert_placer.py]: Moved down hilbert_curve; others are in order (ie _generate_hilbert_chips is called before _place_vertex)

File: pacman/operations/rigged_algorithms/isomorph_check.py (hunk 6 above)
Comment (sara-es) [inline pacman/operations/rigged_algorithms/isomorph_check.py]: No need for additional comment here (going off the "could I explain it in a year's time?" metric)

**Comment by alan-stokes** [comment]
hibert placer, _check_constraints. param types in doc string. object? clearer would be machine vertex, additional placement constraints, set of what?

**Comment by alan-stokes** [comment]
push _hilbert_curve functions into hibertState and make the state variables private and use getters and setters.

**Comment by alan-stokes** [comment]
could ask for comments in hibert state, but meh. Will merge.

**Pull request merged**